text/0000-read-all.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: read_exact and read_full
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's `Write` trait has `write_all`, which is a convenience method that calls
+`write` repeatedly to write an entire buffer. This proposal adds two similar
+convenience methods to the `Read` trait: `read_full` and `read_exact`.
+`read_full` calls `read` repeatedly until the buffer has been filled, EOF has
+been reached, or an error other than `Interrupted` occurs. `read_exact` is
+similar to `read_full`, except that reaching EOF before filling the buffer is
+considered an error.
+
+# Motivation
+
+The `read` method may return fewer bytes than requested, and may fail with an
+`Interrupted` error if a signal is received during the call. This requires
+programs wishing to fill a buffer to call `read` repeatedly in a loop. This is
+a very common need, and it would be nice if this functionality were provided in
+the standard library. Many C and Rust programs have the same need, and solve it
+in the same way. For example, Git has [`read_in_full`][git], which behaves like
+the proposed `read_full`, and the Rust byteorder crate has
+[`read_full`][byteorder], which behaves like the proposed `read_exact`.
+[git]: https://github.com/git/git/blob/16da57c7c6c1fe92b32645202dd19657a89dd67d/wrapper.c#L246
+[byteorder]: https://github.com/BurntSushi/byteorder/blob/2358ace61332e59f596c9006e1344c97295fdf72/src/new.rs#L184
+
+# Detailed design
+
+The following methods will be added to the `Read` trait:
+
+``` rust
+fn read_full(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
@rkjnsn 

> Can you explain this a bit more? Why isn't it composable?

I did indeed miss earlier that `read_full` was doing a loop, but I [agree with @nwin](https://github.com/rust-lang/rfcs/pull/980/files#r32781467) that there are two separable concerns here: ignoring EINTR ignoring short reads. The `read_exact` method as proposed here is doing both, and the `read_full` is kind of also doing both, so there's a lot of overlap between the two. My thoughts about composability stemmed from my interpretation of `read_full` *only* ignoring EINTR, which isn't the case here.

Overall I feel there is too much overlap between these two functions to have both of them. There are many different scenarios which need various forms of loops around `read`, but I don't think we should cater to every last one, only perhaps the most common. I suspect that ignoring EINTR and filling an entire buffer is quite common, which is why I think having `read_exact` is fine, but "fill this buffer as much as possible" does not seem as common to me.